gc/gc-base/src/main/java/org/projectnessie/gc/base/DistributedIdentifyContents.java

```diff
@@ -0,0 +1,392 @@
+ * walking all the references (both dead and live).
+ */
+public class DistributedIdentifyContents implements Serializable {
+
+  private static final long serialVersionUID = -5412153338307931076L;
+  private final SparkSession session;
+  private final Map<String, String> config;
+  private static final Logger LOGGER = LoggerFactory.getLogger(DistributedIdentifyContents.class);
+
+  public DistributedIdentifyContents(SparkSession session, Map<String, String> config) {
+    this.session = session;
+    this.config = config;
+  }
+
+  /**
+   * Builds the client builder; default({@link HttpClientBuilder}) or custom, based on the
+   * configuration provided.
+   *
+   * @param configuration map of client builder configurations.
+   * @return {@link NessieApiV1} object.
+   */
+  public static NessieApiV1 getApi(Map<String, String> configuration) {
+    String clientBuilderClassName =
+        configuration.get(GCConfigConstants.CONF_NESSIE_CLIENT_BUILDER_IMPL);
+    NessieClientBuilder builder;
+    if (clientBuilderClassName == null) {
+      // Use the default HttpClientBuilder
+      builder = HttpClientBuilder.builder();
+    } else {
+      // Use the custom client builder
+      try {
+        builder =
+            (NessieClientBuilder)
+                Class.forName(clientBuilderClassName).getDeclaredConstructor().newInstance();
+      } catch (ClassNotFoundException e) {
+        throw new IllegalArgumentException(
+            String.format(
+                "No custom client builder class found for '%s' ", clientBuilderClassName));
+      } catch (InvocationTargetException
+          | InstantiationException
+          | IllegalAccessException
+          | NoSuchMethodException e) {
+        throw new IllegalArgumentException(
+            String.format("Could not initialize '%s': ", clientBuilderClassName), e);
+      }
+    }
+    return (NessieApiV1) builder.fromConfig(configuration::get).build(NessieApiV1.class);
+  }
+
+  /**
+   * Compute the bloom filter per content id by walking all the live references in a distributed way
+   * using spark.
+   *
+   * @param references list of live references to walk.
+   * @param cutOffTimestampPerRef map of cutoff timestamp per reference.
+   * @param defaultCutOffTimestamp default cutoff timestamp to be used.
+   * @return map of {@link ContentBloomFilter} per content-id.
+   */
+  public Map<String, ContentBloomFilter> getLiveContentsBloomFilters(
+      List<Reference> references,
+      Map<String, Instant> cutOffTimestampPerRef,
+      Instant defaultCutOffTimestamp,
+      Instant deadReferenceCutOffTimeStamp,
+      Map<String, Instant> droppedRefTimeMap,
+      Long totalCommitsInDefaultReference) {
+    List<Map<String, ContentBloomFilter>> bloomFilterMaps =
+        new JavaSparkContext(session.sparkContext())
+            .parallelize(references, getTaskCount(references))
+            .map(
+                reference ->
+                    computeLiveContents(
```

Comment:
Considering the issue w/ the BF siizing I still favor moving the "remote" part to a separate class.
To be clear here: that's for easier maintenance and clear separation of duties - especially for people who are not familiar with the code flow.

Reply:
done. moved to a separate file